entities/src/main/java/org/odk/collect/entities/storage/Entity.kt

```diff
@@ -1,15 +1,50 @@
 package org.odk.collect.entities.storage
 
-data class Entity @JvmOverloads constructor(
-    val list: String,
-    val id: String,
-    val label: String?,
-    val version: Int = 1,
-    val properties: List<Pair<String, String>> = emptyList(),
-    val state: State = State.OFFLINE
-) {
+sealed interface Entity {
+    val list: String
+    val id: String
+    val label: String?
+    val version: Int
+    val properties: List<Pair<String, String>>
+    val state: State
+
+    data class New(
+        override val list: String,
+        override val id: String,
+        override val label: String?,
+        override val version: Int = 1,
+        override val properties: List<Pair<String, String>> = emptyList(),
+        override val state: State = State.OFFLINE
+    ) : Entity
+
+    data class Saved(
+        override val list: String,
+        override val id: String,
+        override val label: String?,
+        override val version: Int = 1,
+        override val properties: List<Pair<String, String>> = emptyList(),
+        override val state: State = State.OFFLINE
+    ) : Entity
+
     enum class State {
         OFFLINE,
         ONLINE
     }
+
+    fun sameAs(entity: Entity): Boolean {
```

Comment:
This is currently just for testing it looks like. It feels like two identical items at different positions in document order should not be considered the same but maybe it doesn't matter for the intended usage.

Is == in Kotlin only going to compare the list references for properties? Maybe that makes my point above moot.

Reply:
> This is currently just for testing it looks like. It feels like two identical items at different positions in document order should not be considered the same but maybe it doesn't matter for the intended usage.

The idea here was to be able to compare entities without the data that gets added by our repo (currently just `index`, but might also be things like a db ID down the line.). We might have actually end up using this in real code I could imagine.

> Is == in Kotlin only going to compare the list references for properties?

Because `Entity` is a `data class` it'll have an intelligent `equals` implementation (and Kotlin calls `equals` rather than doing reference equality with `==`).

Reply:
In the case of Entities, we're (essentially) guaranteed a unique ID so this is fine. In the case of arbitrary secondary instances, I think the index may be important to consider because duplicates are more likely. We can likely come back to it then.